doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,223 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+    Copyright (c) 2018 Nordic Semiconductor ASA
+    SPDX-License-Identifier: Apache-2.0
+.. _codegen_intro:
+
+Introduction
+============
+For some repetitive or parameterized coding tasks, it's convenient to
+use a code generating tool to build C code fragments, instead of writing
+(or editing) that source code by hand. Such a tool can also access CMake build
+parameters and device tree information to generate source code automatically
+tailored and tuned to a specific project configuration.
+
+The Zephyr project supports a code generating tool that processes template files
+and generates output files. It can be used, for example, to generate source code 
+that creates and fills data structures, adapts programming logic, creates
+configuration-specific code fragments, and more.
+
+.. image:: template_flow.png
+   :width: 500px
+   :align: center
+   :alt: Principle
+
+The code generator uses a template engine for python called jinja2. The jinja2
+templates are typically stored in files with the file extention .jinja2 which are
+rendered into other files like c-source files. The documentation for jinja2 can 
+be found here `Jinja2 <http://jinja.pocoo.org/>`_. Please refer to this site for more
+extensive information, the information in the following sections are short
+descriptions of some of the key features in jinja2. 
+
+Jinja2 files are rendered into other files using the cmake extensions functions
+
+  * zephyr_sources_codegen(..)
+  * zephyr_library_sources_codegen(..). 
+
+The generated source files are added to the Zephyr sources. During build the 
+jinja2 files are processed by jinja2 template engine and the generated source
+files are written to the CMake binary directory.
+
+A jinja2 template is written in a domain specific language, but inherited many
+consepts from python, which means that anyone familiar with python should
+learn to write jinja2 templates with ease. 
+
+Data object
+===========
+
+All templates have access to a data object called ``data``. Through this data
+object will the template have access to information like all the elements in
+the .config file, 
+
+.. code-block:: c
+
+	{{ data['config']['CONFIG_BOARD'] }}
+
+devicetree data,
+ 
+ .. code-block:: c
+
+	{% for device in data['devicetree']['devices'] %}
+	    {{ device }}
+	{% endfor %}
+
+
+and runtime information like defines that has been injected using the CODEGEN_DEFINES
+key word in the CMakeLists.txt file.
+
+.. code-block:: c
+
+	{{ data['runtime']['defines']['MY_DEFINE'] }}
```

Comment:
W need to rename "runtime" to something else.

"cmake" maybe? I'm not yet familiar with what it is.

But the term runtime should always be referencing the actual fw running.

Reply:
Runtime refer to the runtime of the code generator script, while cmake would refer the runtime of cmake which may spawn multiple code generation sessions with different code generation runtime data. Maybe straight forward "jinjagen"?